File: src/main/java/com/hyperdondon/blissgemsremake/api/util/GeometryUtil.java
package com.hyperdondon.blissgemsremake.api.util;

import org.bukkit.Location;
import org.bukkit.util.Vector;

public class GeometryUtil {
    //Circular Stuff
    public static Vector getSphericalVector(double r, double yaw, double pitch) {
        double x = r * Math.cos(pitch) * Math.cos(yaw);
        double y = r * Math.sin(pitch);
        double z = r * Math.cos(pitch) * Math.sin(yaw);
        return new Vector(x, y, z);
    }

    public static Vector getSphericalVector(double r, double yaw) {
        return getSphericalVector(r, yaw, 0);
    }

    public static double getParticleAmount(double r, double density) {
        return (2 * Math.PI * r / density);
    }

    public static double getYaw(double r, double iteration, double density) {
        return (iteration * 360) / getParticleAmount(r, density);
    }

    //Directional stuff
    public static Location forwardsBlocks(Location pLocation, double pBlocks) {
        Location clone = pLocation.clone();
        Vector direction = clone.getDirection().normalize();
        return clone.add(direction.multiply(pBlocks));
    }

    public static Location backwardsBlocks(Location pLocation, double pBlocks) {
        Location clone = pLocation.clone();
        Vector direction = clone.getDirection().normalize().multiply(-pBlocks);
        return clone.add(direction);
    }

    public static Location leftBlocks(Location pLocation, double pBlocks) {
        Location clone = pLocation.clone();
        float yaw = clone.getYaw();

        double radians = Math.toRadians(yaw - 90);
        Vector left = new Vector(-Math.sin(radians), 0, Math.cos(radians)).normalize().multiply(pBlocks);
        return clone.add(left);
    }

    public static Location rightBlocks(Location pLocation, double pBlocks) {
        Location clone = pLocation.clone();
        float yaw = clone.getYaw();

        double radians = Math.toRadians(yaw + 90);
        Vector right = new Vector(-Math.sin(radians), 0, Math.cos(radians)).normalize().multiply(pBlocks);
        return clone.add(right);
    }

    public static Location leftBlocksWithPitch(Location pLocation, double pBlocks) {
        Location clone = pLocation.clone();
        clone.setYaw(clone.getYaw() - 90);
        Vector direction = clone.getDirection().normalize().multiply(pBlocks);
        return clone.add(direction);
    }

    public static Location rightBlocksWithPitch(Location pLocation, double pBlocks) {
        Location clone = pLocation.clone();
        clone.setYaw(clone.getYaw() + 90);
        Vector direction = clone.getDirection().normalize().multiply(pBlocks);
        return clone.add(direction);
    }

    public static Location aboveBlocks(Location pLocation, double pBlocks) {
        Location clone = pLocation.clone();
        Vector direction = clone.getDirection(); // Forward
        Vector right = direction.clone().crossProduct(new Vector(0, 1, 0)).normalize(); // Right
        Vector up = right.clone().crossProduct(direction).normalize(); // Local up
        return clone.add(up.multiply(pBlocks));
    }

    public static Location underBlocks(Location pLocation, double pBlocks) {
        Location clone = pLocation.clone();
        Vector direction = clone.getDirection(); // Forward
        Vector right = direction.clone().crossProduct(new Vector(0, 1, 0)).normalize(); // Right
        Vector up = right.clone().crossProduct(direction).normalize(); // Local up
        return clone.subtract(up.multiply(pBlocks));
    }
}

File: src/main/kotlin/com/hyperdondon/blissgemsremake/internal/util/GeometryUtil.kt
package com.hyperdondon.blissgemsremake.internal.util

import org.bukkit.Location
import org.bukkit.util.Vector
import kotlin.math.cos
import kotlin.math.sin

object GeometryUtil {
    //Circular Stuff
    fun getSphericalVector(r: Double, yaw: Double, pitch: Double): Vector {
        val x = r * cos(pitch) * cos(yaw)
        val y = r * sin(pitch)
        val z = r * cos(pitch) * sin(yaw)
        return Vector(x, y, z)
    }

    fun getSphericalVector(r: Double, yaw: Double): Vector {
        return getSphericalVector(r, yaw, 0.0)
    }

    fun getParticleAmount(r: Double, density: Double): Double {
        return (2 * Math.PI * r / density)
    }

    fun getYaw(r: Double, iteration: Double, density: Double): Double {
        return (iteration * 360) / getParticleAmount(r, density)
    }

    //Directional stuff
    fun forwardsBlocks(pLocation: Location, pBlocks: Double): Location {
        val clone = pLocation.clone()
        val direction = clone.direction.normalize()
        return clone.add(direction.multiply(pBlocks))
    }

    fun backwardsBlocks(pLocation: Location, pBlocks: Double): Location {
        val clone = pLocation.clone()
        val direction = clone.direction.normalize().multiply(-pBlocks)
        return clone.add(direction)
    }

    fun leftBlocks(pLocation: Location, pBlocks: Double): Location {
        val clone = pLocation.clone()
        val yaw = clone.yaw

        val radians = Math.toRadians((yaw - 90).toDouble())
        val left = Vector(-sin(radians), 0.0, cos(radians)).normalize().multiply(pBlocks)
        return clone.add(left)
    }

    fun rightBlocks(pLocation: Location, pBlocks: Double): Location {
        val clone = pLocation.clone()
        val yaw = clone.yaw

        val radians = Math.toRadians((yaw + 90).toDouble())
        val right = Vector(-sin(radians), 0.0, cos(radians)).normalize().multiply(pBlocks)
        return clone.add(right)
    }

    fun leftBlocksWithPitch(pLocation: Location, pBlocks: Double): Location {
        val clone = pLocation.clone()
        clone.yaw = clone.yaw - 90
        val direction = clone.direction.normalize().multiply(pBlocks)
        return clone.add(direction)
    }

    fun rightBlocksWithPitch(pLocation: Location, pBlocks: Double): Location {
        val clone = pLocation.clone()
        clone.yaw = clone.yaw + 90
        val direction = clone.direction.normalize().multiply(pBlocks)
        return clone.add(direction)
    }

    fun aboveBlocks(pLocation: Location, pBlocks: Double): Location {
        val clone = pLocation.clone()
        val direction = clone.direction // Forward
        val right = direction.clone().crossProduct(Vector(0, 1, 0)).normalize() // Right
        val up = right.clone().crossProduct(direction).normalize() // Local up
        return clone.add(up.multiply(pBlocks))
    }

    fun underBlocks(pLocation: Location, pBlocks: Double): Location {
        val clone = pLocation.clone()
        val direction = clone.direction // Forward
        val right = direction.clone().crossProduct(Vector(0, 1, 0)).normalize() // Right
        val up = right.clone().crossProduct(direction).normalize() // Local up
        return clone.subtract(up.multiply(pBlocks))
    }
}